motr/io_nw_xfer.c

```diff
@@ -2181,8 +2189,20 @@ static int nw_xfer_req_dispatch(struct nw_xfer_request *xfer)
 			m0_op_io_to_rpc_map(ioo,
 					&irfop->irf_iofop.if_fop.f_item);
 
-			if (rc != 0)
+			if (rc != 0) {
```

Comment:
Please also consider, if the `ioreq_fop_async_submit()` needs to return actually rpc post rc, instead of hardcoded `0`.

```
 672 M0_INTERNAL int ioreq_fop_async_submit(struct m0_io_fop      *iofop,
 673                                        struct m0_rpc_session *session)
 674 {
 675         int                   rc;
 676         struct m0_fop_cob_rw *rwfop;
 677         struct m0_rpc_item   *item;
 678
 679         M0_PRE(iofop != NULL);
 680         M0_PRE(session != NULL);
 681
 682         rwfop = io_rw_get(&iofop->if_fop);
 683         M0_ASSERT(rwfop != NULL);
 684
 685         rc = m0_rpc_bulk_store(&iofop->if_rbulk, session->s_conn,
 686                                rwfop->crw_desc.id_descs,
 687                                &client__buf_bulk_cb);
 688         if (rc != 0)
 689                 goto out;
 690
 691         item = &iofop->if_fop.f_item;
 692         item->ri_session = session;
 693         item->ri_nr_sent_max = M0_RPC_MAX_RETRIES;
 694         item->ri_resend_interval = M0_RPC_RESEND_INTERVAL;
 695         M0_ENTRY("m0_io_fop %p m0_rpc_session %p item = %p", iofop, session,
 696                  item);
 697         rc = m0_rpc_post(item);
 698         M0_LOG(M0_INFO, "IO fops submitted to rpc, rc = %d", rc);
 699
 700         M0_ADDB2_ADD(M0_AVI_CLIENT_BULK_TO_RPC, iofop->if_rbulk.rb_id,
 701                      m0_sm_id_get(&item->ri_sm));
 702         /*
 703          * Ignoring error from m0_rpc_post() so that the subsequent fop
 704          * submission goes on. This is to ensure that the ioreq gets into dgmode
 705          * subsequently without exiting from the healthy mode IO itself.
 706          */
 707         return M0_RC(0);
...

```

Reply:
> Please also consider, if the `ioreq_fop_async_submit()` needs to return actually rpc post rc, instead of hardcoded `0`.
> 
> ```
>  672 M0_INTERNAL int ioreq_fop_async_submit(struct m0_io_fop      *iofop,
>  673                                        struct m0_rpc_session *session)
>  674 {
>  675         int                   rc;
>  676         struct m0_fop_cob_rw *rwfop;
>  677         struct m0_rpc_item   *item;
>  678
>  679         M0_PRE(iofop != NULL);
>  680         M0_PRE(session != NULL);
>  681
>  682         rwfop = io_rw_get(&iofop->if_fop);
>  683         M0_ASSERT(rwfop != NULL);
>  684
>  685         rc = m0_rpc_bulk_store(&iofop->if_rbulk, session->s_conn,
>  686                                rwfop->crw_desc.id_descs,
>  687                                &client__buf_bulk_cb);
>  688         if (rc != 0)
>  689                 goto out;
>  690
>  691         item = &iofop->if_fop.f_item;
>  692         item->ri_session = session;
>  693         item->ri_nr_sent_max = M0_RPC_MAX_RETRIES;
>  694         item->ri_resend_interval = M0_RPC_RESEND_INTERVAL;
>  695         M0_ENTRY("m0_io_fop %p m0_rpc_session %p item = %p", iofop, session,
>  696                  item);
>  697         rc = m0_rpc_post(item);
>  698         M0_LOG(M0_INFO, "IO fops submitted to rpc, rc = %d", rc);
>  699
>  700         M0_ADDB2_ADD(M0_AVI_CLIENT_BULK_TO_RPC, iofop->if_rbulk.rb_id,
>  701                      m0_sm_id_get(&item->ri_sm));
>  702         /*
>  703          * Ignoring error from m0_rpc_post() so that the subsequent fop
>  704          * submission goes on. This is to ensure that the ioreq gets into dgmode
>  705          * subsequently without exiting from the healthy mode IO itself.
>  706          */
>  707         return M0_RC(0);
> ...
> ```

I guess rpc error is captured in ri_error;

Reply:
For rpc error we will have bottom half handling of the fop; but will need to handle same for rpc bulk error; just had a team discussion with Madhav, Yeshpal, Shashank, Yatin on this. We can set the ti_rc = rc in case of bulk error n then again return rc = 0 in dispatch as long single fop is successfully submitted or reported rpc error;